flax/linen/attention.py

```diff
@@ -15,23 +15,42 @@
 """Attention core modules for Flax."""
 
 from functools import partial
-from typing import (Any, Callable, Tuple, Optional)
+from typing import Any, Callable, Tuple, Type, Optional
 
 import jax
 from jax import lax
 from jax import random
 import jax.numpy as jnp
 import numpy as np
+from typing_extensions import Protocol
 
-from flax.linen.linear import default_kernel_init
+from flax.linen.initializers import zeros
 from flax.linen.linear import DenseGeneral
+from flax.linen.linear import _canonicalize_dtypes
+from flax.linen.linear import default_kernel_init
 from flax.linen.module import Module, compact, merge_param
-from flax.linen.initializers import zeros
 
 PRNGKey = Any
-Shape = Tuple[int]
-Dtype = Any
+Shape = Tuple[int, ...]
+InexactDType = Type[jnp.inexact]
 Array = Any
+Initializer = Callable[[PRNGKey, Shape, InexactDType], Array]
```

Comment:
Not for this PR, but it might be useful also for downstream packages to put the definitions of those types in a `flax.typing` or `flax.types` submodule?

Reply:
Yeah, `Initializer` definitely belongs somewhere better.

FYI, I suggested that `Shape` be added to numpy ages ago, and they rightly turned down [the suggestion](https://github.com/numpy/numpy/issues/16657).  I think Stephen's reasoning applies here that the others are probably too short and simple?

Reply:
I added https://github.com/google/jax/pull/9596, which if accepted, would put `Initializer` in `jax.nn` where I think it belongs.

Reply:
By the way, I just found out that MyPy has a crazy bug if you name a file `typing` or `types`: https://github.com/python/mypy/issues/1876